File: apps/web/lib/providers/activepieces.ts
import type { TriageInput, TriageProvider, TriageResult } from "./types";
import { TriageOutput } from "../schema";

/**
 * Full ReAct agent path: POSTs to the Cloud MCP-entry flow's webhook URL
 * (spec 003 `mcp-maintainer-entry`, returnsResponse:true), which runs the
 * agent step with tools + KB + approvals and returns structured JSON.
 *
 * Auth note (verified 2026-09-19): the MCP server at /mcp is OAuth-only
 * ("No API keys to manage") — no static token exists. Webhook triggers
 * are the firewall-friendly path: set AP_FLOW_WEBHOOK_URL after creating
 * the flow in Cloud. Until then: fail closed (501).
 */
export const activepiecesProvider: TriageProvider = {
  name: "activepieces",
  model: process.env.TRIAGE_MODEL || "nvidia/nemotron-3-super-120b-a12b:free",
  async triage(input: TriageInput): Promise<TriageResult> {
    const t0 = Date.now();
    const webhook = (process.env.AP_FLOW_WEBHOOK_URL || "").replace(/\/$/, "");
    if (!webhook) {
      throw Object.assign(
        new Error("Activepieces provider needs AP_FLOW_WEBHOOK_URL (flow webhook from spec 003)."),
        { status: 501, code: "not_wired" }
      );
    }
    const body = JSON.stringify({
      repo: input.repo,
      issue_url: input.issue_url,
      model: (this as TriageProvider).model,
      title: input.title,
      body: input.body,
    });
    // One retry on empty bodies: cold engine runs occasionally return {}
    // before the run completes. Bounded to 2 attempts (agent runs cost).
    let data: unknown = null;
    for (let attempt = 1; attempt <= 2; attempt++) {
      const res = await fetch(webhook, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
      });
      if (!res.ok) {
        throw Object.assign(
          new Error(`MCP-entry flow returned ${res.status}.`),
          { status: 502, code: "provider_error" }
        );
      }
      data = await res.json().catch(() => null);
      const out = (data as { output?: unknown } | null)?.output ?? data;
      if (out && Object.keys(out as object).length > 0) break;
      data = null;
      if (attempt === 1) await new Promise((r) => setTimeout(r, 15000));
    }
    const parsed = TriageOutput.safeParse(
      (data as { output?: unknown } | null)?.output ?? data
    );
    if (!parsed.success) {
      throw Object.assign(
        new Error("Agent returned output outside the contract."),
        { status: 502, code: "provider_error" }
      );
    }
    return {
      output: parsed.data,
      meta: {
        provider: "activepieces",
        model: (this as TriageProvider).model,
        mode: "live-agent",
        latency_ms: Date.now() - t0,
      },
    };
  },
};

File: apps/api/app/providers/activepieces.py
"""Full-agent path via the Cloud MCP-entry flow webhook (spec 003).

Auth note (verified 2026-09-19): the MCP server at /mcp is OAuth-only —
no static token exists, and the API key is rejected there (401). Flow
webhook triggers are the firewall-friendly path: set AP_FLOW_WEBHOOK_URL
after creating the flow in Cloud. Until then: fail closed (501).
"""
import os
import time

import httpx

from .base import ProviderResult, TriageInput, timed
from ..contracts import TriageOutput

NAME = "activepieces"


class ProviderError(Exception):
    def __init__(self, message: str, status: int = 500, code: str = "provider_error"):
        super().__init__(message)
        self.status = status
        self.code = code


class ActivepiecesProvider:
    name = NAME
    model = os.environ.get("TRIAGE_MODEL", "qwen/qwen3.8-27b:free")

    def triage(self, inp: TriageInput) -> ProviderResult:
        from ..repos import KNOWN_REPOS

        t0 = time.time()
        if inp.repo not in KNOWN_REPOS:
            raise ProviderError(
                "Repo not onboarded to the Cloud agent (memory/KB cover "
                "4 seed repos). Use mock/openrouter/openai for any public repo.",
                status=400, code="unknown_repo",
            )
        webhook = os.environ.get("AP_FLOW_WEBHOOK_URL", "").rstrip("/")
        if not webhook:
            raise ProviderError(
                "Activepieces provider needs AP_FLOW_WEBHOOK_URL "
                "(flow webhook from spec 003).",
                status=501, code="not_wired",
            )
        payload: dict = {}
        # One retry on empty bodies: cold engine runs occasionally return {}
        # before the run completes. Bounded to 2 attempts (agent runs cost).
        for attempt in (1, 2):
            try:
                res = httpx.post(
                    webhook,
                    headers={"Content-Type": "application/json"},
                    json={"repo": inp.repo, "issue_url": inp.issue_url,
                          "model": self.model,
                          "title": inp.title, "body": inp.body},
                    timeout=120,
                )
            except httpx.HTTPError as e:
                raise ProviderError(f"Flow webhook unreachable: {e}.") from e
            if res.status_code >= 400:
                raise ProviderError(f"MCP-entry flow returned {res.status_code}.")
            try:
                payload = res.json()
            except ValueError as e:
                raise ProviderError("MCP-entry flow returned non-JSON.") from e
            if payload.get("output") or (
                isinstance(payload, dict) and payload.get("issue_type")
            ):
                break
            time.sleep(15)
        try:
            output = TriageOutput.model_validate(payload.get("output", payload))
        except Exception as e:
            raise ProviderError("Agent returned output outside the contract.") from e
        return timed(NAME, self.model, "live-agent", t0, output)
